src/raw/allocated.rs

```diff
@@ -5,71 +5,78 @@ use core::mem::{transmute, MaybeUninit};
 use core::ops::Range;
 use core::panic::{RefUnwindSafe, UnwindSafe};
 
+// TODO remove once provenance API stabilized
+use sptr::Strict;
+
 use crate::alloc::vec::Vec;
+use crate::backend::rc::Inner;
 use crate::backend::{rc, Backend};
 
 const MASK: usize = super::MASK as usize;
 const TAG: usize = super::TAG_ALLOCATED as usize;
 
-struct TaggedRaw<B: Backend>(*mut (), PhantomData<rc::Raw<Vec<u8>, B>>);
+struct TaggedRawRc<B: Backend>(usize, PhantomData<rc::Raw<Vec<u8>, B>>);
 
-impl<B: Backend> Clone for TaggedRaw<B> {
+impl<B: Backend> Clone for TaggedRawRc<B> {
     #[cfg_attr(coverage_nightly, coverage(off))]
     fn clone(&self) -> Self {
         *self
     }
 }
 
-impl<B: Backend> Copy for TaggedRaw<B> {}
+impl<B: Backend> Copy for TaggedRawRc<B> {}
 
-impl<B: Backend> TaggedRaw<B> {
+impl<B: Backend> TaggedRawRc<B> {
+    /// Constructed a tagged rc from a `rc::Raw`.
     #[inline]
+    #[allow(unstable_name_collisions)]
     fn from(raw: rc::Raw<Vec<u8>, B>) -> Self {
         // SAFETY: add a 2-bit tag to a non-null pointer with the same alignment
         // requirement as usize (typically 4 bytes on 32-bit architectures, and
         // more on 64-bit architectures)
         unsafe {
-            let ptr: *mut () = transmute(raw);
+            let ptr: *mut Inner<Vec<u8>, B> = transmute(raw);
             debug_assert!((ptr as usize) & MASK == 0);
             debug_assert!(!ptr.is_null());
 
-            // Strict provenance API, nightly only, for now just for MIRI
-            #[cfg(miri)]
-            let new_ptr = ptr.map_addr(|addr| addr | TAG);
-
-            #[cfg(not(miri))]
-            let new_ptr = ((ptr as usize) | TAG) as *mut ();
+            // TODO use strict and exposed provenance API once stabilized
+            let addr = ptr.map_addr(|addr| addr | TAG).expose_addr();
```

Comment:
Why does this need expose_addr()? Can't you just use `*mut ()` as before?

Reply:
```suggestion
            let addr = ptr.map_addr(|addr| addr | TAG) as *mut ();
```
Something like this?

Reply:
To be more explicit, from what I following on rust-lang/rust the API will be stabilized shortly and new warning will be emitted.

Reply:
The point of this is to force expose the pointer provenance. Allowing at the end the transmutation to some byte arrays without losing provenance.